Olá estudante!

Me chamo Rafael Meirelles e irei revisar o seu projeto hoje e em eventuais futuras submissões até que ele cumpra todos os requisitos para o aceite.
Conte comigo nessa jornada e não se preocupe se precisar ajustar alguns detalhes, é parte do processo e fundamental para que você exercite os conceitos que vem aprendendo e assim melhore a qualidade dos seus códigos e análises.

Peço que mantenha e não altereos comentários que eu fizer por aqui para que possamos nos localizar posteriormente, ok?

Mais uma coisa, vamos utilizar um código de cores para você entender os meus feedbacks no seu notebook. Funciona assim:

<div class="alert alert-danger">
<strong>Vermelho</strong>

Erro que precisa ser consertado, caso contrário, seu projeto não pode ser aceito

</div>

<div class="alert alert-warning">
<strong>Amarelo</strong>

Alerta de um erro não crítco, mas que pode ser corrigido para melhoria geral no seu código/análise

</div>

<div class="alert alert-success">
<strong>Verde</strong>

Elogios
</div>

<div class="alert alert-info">
<strong>Comentário do estudante</strong>

Use uma caixa azul como essa para eventuais comentários que você gostaria de fazer para mim.
</div>

# Análise do risco de inadimplência dos mutuários

Seu projeto é preparar um relatório para a divisão de empréstimos de um banco. Você precisará descobrir se o estado civil de um cliente e o número de filhos têm impacto sobre se ele deixará de pagar um empréstimo. O banco já tem alguns dados sobre a capacidade de crédito dos clientes.

Seu relatório será considerado ao criar uma **pontuação de crédito** de um cliente em potencial. A **contagem de crédito** é usada para avaliar a capacidade de um devedor em potencial de pagar seu empréstimo.



## Abra o arquivo de dados e veja as informações gerais.


In [1]:
# Carregando todas as bibliotecas                 
import pandas as pd
# Carregue os dados
data = pd.read_csv('dados/credit_scoring_eng.csv')

## Tarefa 1. Exploração de dados

**Descrição dos dados**
- `children` - o número de crianças na família
- `days_employed` - experiência de trabalho em dias
- `dob_years` - idade do cliente em anos
- `education` - educação do cliente
- `education_id` - identificador de educação
- `family_status` - estado civil do cliente
- `family_status_id` - identificador de estado civil
- `gender` - gênero do cliente
- `income_type` - tipo de emprego
- `debt` - havia alguma dívida no pagamento do empréstimo
- `total_income` - renda mensal
- `purpose` - o objetivo de obter um empréstimo


In [2]:
# Vamos ver quantas linhas e colunas nosso conjunto de dados tem          
data.shape

(21525, 12)

In [3]:
# vamos exibir as primeiras N linhas
data.head(15)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42    bachelor's degree             0   
1          1   -4024.803754         36  secondary education             1   
2          0   -5623.422610         33  Secondary Education             1   
3          3   -4124.747207         32  secondary education             1   
4          0  340266.072047         53  secondary education             1   
5          0    -926.185831         27    bachelor's degree             0   
6          0   -2879.202052         43    bachelor's degree             0   
7          0    -152.779569         50  SECONDARY EDUCATION             1   
8          2   -6929.865299         35    BACHELOR'S DEGREE             0   
9          0   -2188.756445         41  secondary education             1   
10         2   -4171.483647         36    bachelor's degree             0   
11         0    -792.701887         40  secondary education             1   
12         0            NaN         65  secondary education             1   
13         0   -1846.641941         54         some college             2   
14         0   -1844.956182         56    bachelor's degree             0   

        family_status  family_status_id gender income_type  debt  \
0             married                 0      F    employee     0   
1             married                 0      F    employee     0   
2             married                 0      M    employee     0   
3             married                 0      M    employee     0   
4   civil partnership                 1      F     retiree     0   
5   civil partnership                 1      M    business     0   
6             married                 0      F    business     0   
7             married                 0      M    employee     0   
8   civil partnership                 1      F    employee     0   
9             married                 0      M    employee     0   
10            married                 0      M    business     0   
11            married                 0      F    employee     0   
12  civil partnership                 1      M     retiree     0   
13            married                 0      F    employee     0   
14  civil partnership                 1      F    business     1   

    total_income                              purpose  
0      40620.102                purchase of the house  
1      17932.802                         car purchase  
2      23341.752                purchase of the house  
3      42820.568              supplementary education  
4      25378.572                    to have a wedding  
5      40922.170                purchase of the house  
6      38484.156                 housing transactions  
7      21731.829                            education  
8      15337.093                     having a wedding  
9      23108.150  purchase of the house for my family  
10     18230.959                      buy real estate  
11     12331.077           buy commercial real estate  
12           NaN                    to have a wedding  
13     20873.317                         car purchase  
14     26420.466          buy residential real estate

Realizando a verificação nas primeiras 15 linhas, foi identificado: 

1-) coluna ‘days_employed - Possui números negativos para dias trabalhados o que não faz sentido para dias que pela lógica deveriam ser positivo uma vez que já foi contabilizado o dia trabalhado’, será necessário maior investigação e uma possível alteração nos dados de negativos para positivos.

2-) coluna ‘education’ - Existem diferenças de letras maiúsculas e minúsculas o que pode impactar uma possível agregação para análise dos dados, será necessário alterações para padronizar dados.

3-) colunas days_employed‘ e ‘income_type’ possuem dados ausentes, será necessário realizar a tratativa para estas colunas.


In [4]:
# Obter informações sobre dados   OK
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [5]:
data.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000000      0.000000   32549.611000  
max            4.000000      1.000000  362496.645000

<div class="alert alert-warning">
<strong>Comentário do revisor v1</strong>

Lembre-se que o describe() pode te ajudar a identificar problemas e características dos dados.
</div>

As colunas ‘days_employed‘ e ‘income_type’ possuem dados ausentes simultaneamente, será preciso uma abertura nestes campos para uma melhor verificação de correlação dos dados referente a um possível motivo da ausência dos dados.


In [6]:
# Vejamos a tabela filtrada com valores ausentes na primeira coluna com dados ausentes   OK
print(data['days_employed'].isna().sum())

2174


In [7]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [8]:
nulos = data.loc[(data['days_employed'].isnull()) & (data['total_income'].isnull())]
nulos

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

Analisando as colunas 'days employed' e  'income type' elas possuem simetricamente dados ausentes em ambas, podemos levantar mais hipóteses relacionadas a este tópico, realizando uma abertura de dados sobre os campos, family status, income type, education, por exemplo para tentar identificar possíveis informações que ajudem a basear as argumentações para ausência de dados.           


<div class="alert alert-success">
<strong>Comentário do revisor v1</strong>

Bom trabalho com essa identificação.
</div>

In [9]:
# Vamos aplicar várias condições para filtrar dados e observar o número de linhas na tabela filtrada.
df_nulos = nulos.groupby('family_status')['income_type'].count()
df_nulos

family_status
civil partnership     442
divorced              112
married              1237
unmarried             288
widow / widower        95
Name: income_type, dtype: int64

In [10]:
df_nulos = nulos.groupby('income_type')['income_type'].count()
df_nulos

income_type
business          508
civil servant     147
employee         1105
entrepreneur        1
retiree           413
Name: income_type, dtype: int64

In [11]:
df_nulos = nulos.groupby('education')['income_type'].count()
df_nulos

education
BACHELOR'S DEGREE        23
Bachelor's Degree        25
PRIMARY EDUCATION         1
Primary Education         1
SECONDARY EDUCATION      67
SOME COLLEGE              7
Secondary Education      65
Some College              7
bachelor's degree       496
primary education        19
secondary education    1408
some college             55
Name: income_type, dtype: int64

In [12]:
df_nulos = nulos.groupby('debt')['income_type'].count()
df_nulos

debt
0    2004
1     170
Name: income_type, dtype: int64

In [13]:
df_nulos = nulos.groupby('children')['income_type'].count()
df_nulos

children
-1        3
 0     1439
 1      475
 2      204
 3       36
 4        7
 5        1
 20       9
Name: income_type, dtype: int64

Os valores ausentes em questão são simétricos entre os campos days employed e total income, e representam um percentual de 0,10% do total de dados, um valor que não é relevantemente alto para prejudicar a análise de crédito final.

Será nescessário uma maior verificação nas colunas com dados ausentes para evitar a exclusão de dados, vamos manter realizando uma substituição com a média ou mediana, que será decidida a ação nas próximas análises. 

Realizando uma maior abertura nos dados referente a status family, education, total income, children  não possui uma característica evidente da ausência de dados e sim algumas normalizações serão precisas como exemplo a coluna 'children', consta dados negativos de filhos mostrando uma informação incorreta. 

<div class="alert alert-warning">
<strong>Comentário do revisor v1</strong>

Me desculpa, não entendi muito bem esse passo aqui. Você optou por remover das análises dessa sessão mas não do notebook como um todo, é isso?
</div>

<div class="alert alert-info">
<strong>Comentário do estudante</strong>

Realizei a correção usei uma interpretação anterior e esqueci de realizar a alteração no texto agora esta feito
</div>

In [14]:
# Vamos investigar clientes que não possuem dados sobre as características identificadas e a coluna com os valores ausentes
df_nulos = nulos.groupby('income_type')['income_type'].count()
df_nulos

income_type
business          508
civil servant     147
employee         1105
entrepreneur        1
retiree           413
Name: income_type, dtype: int64

In [15]:
data.groupby('income_type')['income_type'].count()

income_type
business                        5085
civil servant                   1459
employee                       11119
entrepreneur                       2
paternity / maternity leave        1
retiree                         3856
student                            1
unemployed                         2
Name: income_type, dtype: int64

In [16]:
# Contagem de dados nulos x ausentes e seu percentual sobre o total
total_nulos = nulos.shape[0]
print('Total nulos, ',total_nulos)
total_dados = data.shape[0]
print('Total de dados no geral, ',total_dados)
percentual  = round((total_nulos / total_dados),2)
print('Temos um total percentual de dados ausentes na base de ',percentual,'%')

Total nulos,  2174
Total de dados no geral,  21525
Temos um total percentual de dados ausentes na base de  0.1 %


Verificando a coluna income_type onde possui os dados classificados sobre status ativo e/ou inativo, comparando com o total geral podemos perceber que a volumetria é semelhante não apresentando nenhum desvio representativo, pois o valor geral de dados nulos é de 0,10%, como por exemplo a lógica de ter mais valores inseridos como desempregados como valores não declarados e sim são aleatórios neste caso temos mais casos de clientes empregados e mesmo assim não foi declarado em sistema o valor recebido e/ou pode ter ocorrido algo no sistema.

In [17]:
# Verificar a distribuição em todo o conjunto de dados  (inserir aqui as demais colunas)
nulos.groupby('education')['income_type'].count()

education
BACHELOR'S DEGREE        23
Bachelor's Degree        25
PRIMARY EDUCATION         1
Primary Education         1
SECONDARY EDUCATION      67
SOME COLLEGE              7
Secondary Education      65
Some College              7
bachelor's degree       496
primary education        19
secondary education    1408
some college             55
Name: income_type, dtype: int64

In [18]:
nulos.groupby('family_status')['income_type'].count()

family_status
civil partnership     442
divorced              112
married              1237
unmarried             288
widow / widower        95
Name: income_type, dtype: int64

In [19]:
nulos.groupby('gender')['income_type'].count()

gender
F    1484
M     690
Name: income_type, dtype: int64

Sim, a distribuição é semelhante, não demonstrando desvio significativo para embasar a ausência de dados de forma categórica.

Podemos correlacionar a hipótese de ausência de dados, uma vez que a proporção de dados ausentes e preenchidos são semelhantes ao fator do formulário e/ou sistema captura de informação por exemplo não ser um dado “obrigatório” ou erro na extração de dados, são duas possíveis razões para este problema.


**Conclusões**


Não foi encontrado nenhum padrão para os valores ausentes, para os valores ausentes serão tratados com dados semelhentas de analises de mediana, por ser uma pequena quantia, referente ao todo somente os dados ausentes na coluna débito que possuem dívidas serão considerados.

Para os próximos passos serão realizados:

- exclusão de dados duplicados
- correção da formatação escrita coluna education
- correção de valores negativos na coluna days_employed
- tratativa de dados ausentes em clientes com dívidas 



<div class="alert alert-success">
<strong>Comentário do revisor v1</strong>

Vendo a sua conclusão eu concordo com as suas decisões! Só aquele comentário de cima que me confundiu um pouco por isso optei por mantê-lo lá pra te dar a chance de reescrever.

Mas excelente trabalho nos checks e na conclusão.
</div>

## Transformação de dados


In [20]:
data.head(5)

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

In [21]:
# Vamos ver todos os valores na coluna de educação para verificar se e quais grafias precisarão ser corrigidas
data['education'].value_counts()

secondary education    13750
bachelor's degree       4718
SECONDARY EDUCATION      772
Secondary Education      711
some college             668
BACHELOR'S DEGREE        274
Bachelor's Degree        268
primary education        250
Some College              47
SOME COLLEGE              29
PRIMARY EDUCATION         17
Primary Education         15
graduate degree            4
Graduate Degree            1
GRADUATE DEGREE            1
Name: education, dtype: int64

In [22]:
# Corrija os registros, se necessário
data['education'] = data['education'].str.lower()

In [23]:
# Verificando todos os valores na coluna para ter certeza de que os corrigimos
data['education'].unique()

array(["bachelor's degree", 'secondary education', 'some college',
       'primary education', 'graduate degree'], dtype=object)

<div class="alert alert-success">
<strong>Comentário do revisor v1</strong>

Perfeito!
</div>

In [24]:
# Vamos ver a distribuição de valores na coluna `children`
data['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

In [25]:
age_error = (76 + 47) / data['children'].shape[0]
age_error
print(f'Erro dentro do percentual de {age_error:.2%}, sobre dados de idade com casais com 20 filhos e dados negativos')

Erro dentro do percentual de 0.57%, sobre dados de idade com casais com 20 filhos e dados negativos


In [26]:
# corrigindo os dados 
data = data.loc[ (data['children'] != -1) & (data['children'] != 20) ]
data

children  days_employed  dob_years            education  education_id  \
0             1   -8437.673028         42    bachelor's degree             0   
1             1   -4024.803754         36  secondary education             1   
2             0   -5623.422610         33  secondary education             1   
3             3   -4124.747207         32  secondary education             1   
4             0  340266.072047         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1   -4529.316663         43  secondary education             1   
21521         0  343937.404131         67  secondary education             1   
21522         1   -2113.346888         38  secondary education             1   
21523         3   -3112.481705         38  secondary education             1   
21524         2   -1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                  purpose  
0         40620.102    purchase of the house  
1         17932.802             car purchase  
2         23341.752    purchase of the house  
3         42820.568  supplementary education  
4         25378.572        to have a wedding  
...             ...                      ...  
21520     35966.698     housing transactions  
21521     24959.969        purchase of a car  
21522     14347.610                 property  
21523     39054.888        buying my own car  
21524     13127.587             to buy a car  

[21402 rows x 12 columns]

In [27]:
# Verificar a coluna `children` novamente para ter certeza de que está tudo corrigido
data['children'].value_counts()

0    14149
1     4818
2     2055
3      330
4       41
5        9
Name: children, dtype: int64

In [28]:
data['purpose']

0          purchase of the house
1                   car purchase
2          purchase of the house
3        supplementary education
4              to have a wedding
                  ...           
21520       housing transactions
21521          purchase of a car
21522                   property
21523          buying my own car
21524               to buy a car
Name: purpose, Length: 21402, dtype: object

<div class="alert alert-warning">
<strong>Comentário do revisor v1</strong>

Bom trabalho com a identificação do problema, estou de acordo, no entanto, temos informações para tomar uma decisão melhor do que simplesmente eliminar os dados errados.
</div>

<div class="alert alert-info">
<strong>Comentário do estudante</strong>

Realizei a correção da informação de exclusão de dados 
</div>

In [29]:
# Encontre dados problemáticos em `days_employed`, se existirem, e calcule a porcentagem
data['days_employed'].unique()

array([-8437.67302776, -4024.80375385, -5623.42261023, ...,
       -2113.3468877 , -3112.4817052 , -1984.50758853])

In [30]:
# Aborde os valores problemáticos, se existirem
# alterando para positivo 
data['days_employed'] = abs(data['days_employed'])

In [31]:
# Verifique o resultado - certifique-se de que está corrigido
data.head(3)

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   

  family_status  family_status_id gender income_type  debt  total_income  \
0       married                 0      F    employee     0     40620.102   
1       married                 0      F    employee     0     17932.802   
2       married                 0      M    employee     0     23341.752   

                 purpose  
0  purchase of the house  
1           car purchase  
2  purchase of the house

In [32]:
# verificando após a correção a distribuição da coluna 'days_employed'
# Verificado que os dados da variável 'days_employed' possui dados divergentes que são a quantidade de dias trabalhados maior que a idade máxima
data.describe()

children  days_employed     dob_years  education_id  \
count  21402.000000   19240.000000  21402.000000  21402.000000   
mean       0.473180   67027.691459     43.300206      0.817120   
std        0.751837  139130.846446     12.579055      0.548741   
min        0.000000      24.141633      0.000000      0.000000   
25%        0.000000     927.984311     33.000000      1.000000   
50%        0.000000    2195.251592     42.000000      1.000000   
75%        1.000000    5556.372075     53.000000      1.000000   
max        5.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21402.000000  21402.000000   19240.000000  
mean           0.973133      0.080927   26791.806362  
std            1.420536      0.272729   16496.819745  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16480.020250  
50%            0.000000      0.000000   23203.328000  
75%            1.000000      0.000000   32551.144000  
max            4.000000      1.000000  362496.645000

In [33]:
# Realizando o cálculo de quantos dados incorretos e seu percentual na coluna 'days_employed'
employ_max_year = 27393.75
error_days = data[data['days_employed'] > employ_max_year]
print('Total de dados incorretos na coluna days_employed:' ,error_days.shape[0])
total_erro = round((error_days.shape[0] / data.shape[0]),2)
print('Percentual de daddos incorretos na coluna days_employed comparando com o total: ',total_erro)

Total de dados incorretos na coluna days_employed: 3431
Percentual de daddos incorretos na coluna days_employed comparando com o total:  0.16


In [34]:
# Preparando os dados da coluna para a realização do laço for
adj_days_employed = data['days_employed']
adj_days_employed.head(2)

0    8437.673028
1    4024.803754
Name: days_employed, dtype: float64

In [35]:
# realizando laço for para a substituição dos dados incorretos da coluna 'days_employed', que será realizada pela mediana
for i in adj_days_employed:
    if i > employ_max_year:
        data['days_employed'] = data['days_employed'].replace(i,employ_max_year)   
    

In [36]:
# Verificando se foi realizada a alteração pela mediana 
data.describe()

children  days_employed     dob_years  education_id  \
count  21402.000000   19240.000000  21402.000000  21402.000000   
mean       0.473180    6818.964252     43.300206      0.817120   
std        0.751837    9810.271815     12.579055      0.548741   
min        0.000000      24.141633      0.000000      0.000000   
25%        0.000000     927.984311     33.000000      1.000000   
50%        0.000000    2195.251592     42.000000      1.000000   
75%        1.000000    5556.372075     53.000000      1.000000   
max        5.000000   27393.750000     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21402.000000  21402.000000   19240.000000  
mean           0.973133      0.080927   26791.806362  
std            1.420536      0.272729   16496.819745  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16480.020250  
50%            0.000000      0.000000   23203.328000  
75%            1.000000      0.000000   32551.144000  
max            4.000000      1.000000  362496.645000

<div class="alert alert-danger">
<strong>Comentário do revisor v1</strong>

Bom trabalho com os valores negativos, mas existem outros problemas com essa variável que o describe te ajudaria aidentificar.
</div>

In [37]:
# Verifique o `dob_years` para valores suspeitos e conte a porcentagem
data['dob_years'].value_counts()

35    614
40    603
41    603
34    597
38    595
42    592
33    577
39    572
31    556
36    553
44    543
29    543
30    536
48    536
37    531
43    510
50    509
32    506
49    505
28    501
45    494
27    490
52    483
56    482
47    480
54    476
46    469
58    461
57    457
53    457
51    446
59    441
55    441
26    406
60    376
25    356
61    353
62    351
63    268
64    263
24    263
23    252
65    194
22    183
66    183
67    167
21    110
0     100
68     99
69     83
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

Existe 101 idades cadastradas no valor de 0, para este caso iremos usar uma média para substitur.

In [38]:
# Resolva os problemas na coluna `dob_years`, se existirem
# existem 101 cadastros com idade de 0 anos 
age_avg = data['dob_years'].mean()
age_avg
data['dob_years'] = data['dob_years'].replace(0,43)

In [39]:
# Verifique o resultado - certifique-se de que está corrigido
data['dob_years'].value_counts()

35    614
43    610
41    603
40    603
34    597
38    595
42    592
33    577
39    572
31    556
36    553
29    543
44    543
48    536
30    536
37    531
50    509
32    506
49    505
28    501
45    494
27    490
52    483
56    482
47    480
54    476
46    469
58    461
53    457
57    457
51    446
59    441
55    441
26    406
60    376
25    356
61    353
62    351
63    268
64    263
24    263
23    252
65    194
66    183
22    183
67    167
21    110
68     99
69     83
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

In [40]:
# Vamos ver os valores da coluna
data['family_status'].value_counts()

married              12302
civil partnership     4160
unmarried             2799
divorced              1189
widow / widower        952
Name: family_status, dtype: int64

In [41]:
# Vamos ver os valores na coluna
data['gender'].value_counts()

F      14154
M       7247
XNA        1
Name: gender, dtype: int64

Existe somente um dado que não foi cadastrado entre os gêneros feminino e masculino, para este caso iremos excluir da base de dados. 

<div class="alert alert-success">
<strong>Comentário do revisor v1</strong>

Nesse caso ok excluir!
</div>

In [42]:
# Aborde os valores problemáticos, se existirem
data = data.loc[ (data['gender'] != 'XNA') ]

In [43]:
# Verifique o resultado - certifique-se de que está corrigido
data['gender'].value_counts()

F    14154
M     7247
Name: gender, dtype: int64

In [44]:
# Vamos ver os valores na coluna
data['income_type'].value_counts()

employee                       11050
business                        5053
retiree                         3839
civil servant                   1453
unemployed                         2
entrepreneur                       2
student                            1
paternity / maternity leave        1
Name: income_type, dtype: int64

In [45]:
# Verificar duplicatas
data.duplicated().sum()

71

In [46]:
# Aborde as duplicatas, se existirem
duplicated_data = data[data.duplicated()]
duplicated_data

children  days_employed  dob_years            education  education_id  \
2849          0            NaN         41  secondary education             1   
3290          0            NaN         58  secondary education             1   
4182          1            NaN         34    bachelor's degree             0   
4851          0            NaN         60  secondary education             1   
5557          0            NaN         58  secondary education             1   
...         ...            ...        ...                  ...           ...   
20702         0            NaN         64  secondary education             1   
21032         0            NaN         60  secondary education             1   
21132         0            NaN         47  secondary education             1   
21281         1            NaN         30    bachelor's degree             0   
21415         0            NaN         54  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
2849             married                 0      F    employee     0   
3290   civil partnership                 1      F     retiree     0   
4182   civil partnership                 1      F    employee     0   
4851   civil partnership                 1      F     retiree     0   
5557   civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
20702            married                 0      F     retiree     0   
21032            married                 0      F     retiree     0   
21132            married                 0      F    employee     0   
21281            married                 0      F    employee     0   
21415            married                 0      F     retiree     0   

       total_income                              purpose  
2849            NaN  purchase of the house for my family  
3290            NaN                    to have a wedding  
4182            NaN                     wedding ceremony  
4851            NaN                     wedding ceremony  
5557            NaN                    to have a wedding  
...             ...                                  ...  
20702           NaN              supplementary education  
21032           NaN                   to become educated  
21132           NaN                   housing renovation  
21281           NaN           buy commercial real estate  
21415           NaN                 housing transactions  

[71 rows x 12 columns]

In [47]:
data = data.drop_duplicates()

<div class="alert alert-warning">
<strong>Comentário do revisor v1</strong>

Lembre-se de usar o reset_index() nesses casos
</div>

<div class="alert alert-info">
<strong>Comentário do estudante</strong>

Obrigado por avisar realizado a inserção do comando reset
</div>

In [48]:
# Última verificação se temos duplicatas
data.duplicated().sum()

0

In [49]:
#Verifique o tamanho do conjunto de dados que você tem agora após suas primeiras manipulações com ele
data.shape

(21330, 12)

Dentro deste conjunto de dados foram realizadas as ações de remoção de duplicadas, correção de valores e um checkup em todas colunas de relevancia para contabilização da análise de dados.

# Trabalhando com valores ausentes

In [50]:
# Encontre os dicionários
dic_edu = data.loc[:,['education_id','education']].drop_duplicates().reset_index(drop=True)
dic_edu

education_id            education
0             0    bachelor's degree
1             1  secondary education
2             2         some college
3             3    primary education
4             4      graduate degree

In [51]:
dic_fam = data.loc[:,['family_status_id','family_status']].drop_duplicates().reset_index(drop=True)
dic_fam

family_status_id      family_status
0                 0            married
1                 1  civil partnership
2                 2    widow / widower
3                 3           divorced
4                 4          unmarried

In [52]:
data_debt = data.loc[(data['debt'] == 1 ) & (data['total_income'].isnull())]
data_debt.shape[0]

170

In [53]:
data_debt.head(2)

children  days_employed  dob_years            education  education_id  \
55          0            NaN         54  secondary education             1   
181         0            NaN         26  secondary education             1   

         family_status  family_status_id gender income_type  debt  \
55   civil partnership                 1      F     retiree     1   
181  civil partnership                 1      F    business     1   

     total_income                              purpose  
55            NaN                    to have a wedding  
181           NaN  purchase of the house for my family

### Restaurar valores ausentes em `total_income`

In [54]:
data_nulos = nulos.groupby('income_type')['income_type'].count()
data_nulos

income_type
business          508
civil servant     147
employee         1105
entrepreneur        1
retiree           413
Name: income_type, dtype: int64

Podemos percerber que a maior quantidade de valores ausentes são nas colunas employee, business, retiree, para estas colunas será necessário uma abertura para investigar se dados como tipo de emprego e/ou outras váriaáveis possam interferir e realizar um cálculo dentro de outras colunas já preenchidas para completar estas ausentes e evitar a exclusão dos dados.           

### Verificando dados não ausentes

In [55]:
# Crie uma tabela sem valores ausentes e exiba algumas de suas linhas para garantir que ela fique boa
data_nao_ausente = data.loc[data['days_employed'].isnull() != True]
data_nao_ausente.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

In [56]:
# Veja os valores médios de renda com base em seus fatores identificados
print('O valor médio geral da coluna total_income é',round(data_nao_ausente['total_income'].mean(),2))

O valor médio geral da coluna total_income é 26791.5


In [57]:
# Veja os valores medianos de renda com base em seus fatores identificados
print('O valor da mediana geral da coluna total_income é',round(data_nao_ausente['total_income'].median(),2))

O valor da mediana geral da coluna total_income é 23202.87


In [58]:
data_nao_ausente['total_income'].median()

23202.87

<div class="alert alert-danger">
<strong>Comentário do revisor v1</strong>

O comando acima deu erro e não consegui continuar a análise
</div>

Abaixo tem foi criada uma função que realizar uma categorização de dados referente a idade para identificar um grupo de clientes que são jovens, adultos e aposentados.

In [59]:
# o conjunto gerais de idades para realizar futuras categorizações com esta fonte de dados 
data['dob_years'].value_counts()

35    613
43    609
41    601
40    601
34    595
38    594
42    591
33    577
39    571
31    555
36    552
29    542
44    541
48    534
30    533
37    530
50    508
49    505
32    505
28    501
45    493
27    490
52    483
56    478
47    477
54    473
46    466
53    457
58    454
57    453
51    444
55    441
59    440
26    406
60    373
25    356
61    352
62    347
63    268
24    262
64    258
23    250
65    193
22    183
66    182
67    167
21    110
68     99
69     83
70     65
71     56
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

In [60]:
# Vamos escrever uma função que calcule a categoria de idade
def grupo_idade(idade):    
    if idade <= 20:
        return 'Jovem_Aprendiz'
    if idade <= 65:
        return 'Profissionais_Adultos'
    return 'Aposentados'  

In [61]:
# Teste se a função funciona
print(grupo_idade(30))  

Profissionais_Adultos


In [62]:
# Criar coluna nova com base na função
data['Categoria_Idade'] = data['dob_years'].apply(grupo_idade)
data

children  days_employed  dob_years            education  education_id  \
0             1    8437.673028         42    bachelor's degree             0   
1             1    4024.803754         36  secondary education             1   
2             0    5623.422610         33  secondary education             1   
3             3    4124.747207         32  secondary education             1   
4             0   27393.750000         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1    4529.316663         43  secondary education             1   
21521         0   27393.750000         67  secondary education             1   
21522         1    2113.346888         38  secondary education             1   
21523         3    3112.481705         38  secondary education             1   
21524         2    1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                  purpose        Categoria_Idade  
0         40620.102    purchase of the house  Profissionais_Adultos  
1         17932.802             car purchase  Profissionais_Adultos  
2         23341.752    purchase of the house  Profissionais_Adultos  
3         42820.568  supplementary education  Profissionais_Adultos  
4         25378.572        to have a wedding  Profissionais_Adultos  
...             ...                      ...                    ...  
21520     35966.698     housing transactions  Profissionais_Adultos  
21521     24959.969        purchase of a car            Aposentados  
21522     14347.610                 property  Profissionais_Adultos  
21523     39054.888        buying my own car  Profissionais_Adultos  
21524     13127.587             to buy a car  Profissionais_Adultos  

[21330 rows x 13 columns]

In [63]:
# Verificar com os valores estão na nova coluna
data['Categoria_Idade'].value_counts()

Profissionais_Adultos    20565
Aposentados                700
Jovem_Aprendiz              65
Name: Categoria_Idade, dtype: int64

Com a verificação geral das colunas com dados ausentes para  a coluna total income iremos usar como cálculo o valor da mediana por se tratar de uma calculo que não outliers não irão impactar, referente a coluna days_employed usaremos também a mediana.


In [64]:
# Gerando tabela sem ausentes para realizar estudo de cálculo para as substituições. 
data_sem_ausente = data.loc[~data['days_employed'].isnull()]

In [65]:
#validando se não temos nenhum valor ausente
data_sem_ausente.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
Categoria_Idade     0
dtype: int64

In [66]:
data_sem_ausente.head(2)

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   

  family_status  family_status_id gender income_type  debt  total_income  \
0       married                 0      F    employee     0     40620.102   
1       married                 0      F    employee     0     17932.802   

                 purpose        Categoria_Idade  
0  purchase of the house  Profissionais_Adultos  
1           car purchase  Profissionais_Adultos

In [67]:
# Veja os valores média de renda com base nos dados analisados
# Verificando renda por classificação profissional, genero, status familia

data_pivot = data_sem_ausente.pivot_table(index= ['income_type','gender'],
                                        columns= 'family_status',
                                        values = 'total_income',
                                        aggfunc = 'median')
data_pivot

family_status                       civil partnership    divorced     married  \
income_type                 gender                                              
business                    F              25279.0520  27060.1570  25798.2270   
                            M              31010.1745  30945.7725  31920.6510   
civil servant               F              22645.2030  22967.2795  21524.2720   
                            M              31623.4095  33318.1010  28998.0065   
employee                    F              21508.0030  22205.3930  20400.3120   
                            M              25546.6040  24967.5100  26633.3320   
entrepreneur                F              79866.1030         NaN         NaN   
paternity / maternity leave F                     NaN         NaN   8612.6610   
retiree                     F              18242.0645  19215.4410  18285.4005   
                            M              20877.6770  21416.5410  21262.3100   
student                     M                     NaN         NaN         NaN   
unemployed                  F              32435.6020         NaN         NaN   
                            M                     NaN         NaN   9593.1190   

family_status                        unmarried  widow / widower  
income_type                 gender                               
business                    F       25907.9795       23208.7975  
                            M       30341.8220       27597.9835  
civil servant               F       23078.5040       20621.2900  
                            M       29002.9830       42925.5380  
employee                    F       20876.5340       22199.2560  
                            M       23972.9590       23433.5740  
entrepreneur                F              NaN              NaN  
paternity / maternity leave F              NaN              NaN  
retiree                     F       19011.0475       19128.9470  
                            M       19434.4580       18065.6980  
student                     M       15712.2600              NaN  
unemployed                  F              NaN              NaN  
                            M              NaN              NaN

In [68]:
# Verificando a descrição geral da renda geral do grupo
data_sem_ausente['total_income'].describe()

count     19239.000000
mean      26791.503175
std       16497.194890
min        3306.762000
25%       16479.999500
50%       23202.870000
75%       32549.611000
max      362496.645000
Name: total_income, dtype: float64

In [69]:
#  Escreva uma função que usaremos para preencher os valores ausentes
def fill_absent(family_status,income_type,gender):
    try:
        return(df_pivot[family_status][income_type][gender])
    except:
        return 'erro'

In [70]:
# Verifique se funciona
fill_absent('divorced','employee','M')

'erro'

In [71]:
# aplique em todas as linhas 
data['mediana_income'] = data.apply(lambda row: fill_absent(row['family_status'],row['income_type'],row['gender']), axis=1)

In [72]:
data.head(5)

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0   27393.750000         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose        Categoria_Idade mediana_income  
0    purchase of the house  Profissionais_Adultos           erro  
1             car purchase  Profissionais_Adultos           erro  
2    purchase of the house  Profissionais_Adultos           erro  
3  supplementary education  Profissionais_Adultos           erro  
4        to have a wedding  Profissionais_Adultos           erro

In [73]:
data['total_income'] = data['total_income'].fillna(data['mediana_income'])

In [74]:
#Verificando se erros persistem 
data[data['total_income'] == 'erro']

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  secondary education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    bachelor's degree             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

      total_income                   purpose        Categoria_Idade  \
12            erro         to have a wedding  Profissionais_Adultos   
26            erro                 education  Profissionais_Adultos   
29            erro    building a real estate  Profissionais_Adultos   
41            erro  second-hand car purchase  Profissionais_Adultos   
55            erro         to have a wedding  Profissionais_Adultos   
...            ...                       ...                    ...   
21489         erro         purchase of a car  Profissionais_Adultos   
21495         erro          wedding ceremony  Profissionais_Adultos   
21497         erro       building a property  Profissionais_Adultos   
21502         erro    building a real estate  Profissionais_Adultos   
21510         erro              car purchase  Profissionais_Adultos   

      mediana_income  
12              erro  
26              erro  
29              erro  
41              erro  
55              erro  
...              ...  
21489           erro  
21495           erro  
21497           erro  
21502           erro  
21510           erro  

[2091 rows x 14 columns]

In [75]:
# inserindo valor para a tratativa do erro, no caso a mediana geral
data['total_income'] = data['total_income'].replace('erro',26791.503175)

In [76]:
# Verificando se o tipo de dados estão corretos
data['total_income'].dtype

dtype('float64')

In [77]:
#verificando se ainda conta algum dado ausente na coluna 
data['total_income'].isnull().sum()

0

<div class="alert alert-success">
<strong>Comentários do revisor - v1</strong>

Bom trabalho com a função.
    
</div>

###  Restaurar valores em `days_employed`

In [78]:
data_sem_ausente.head(5)

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0   27393.750000         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose        Categoria_Idade  
0    purchase of the house  Profissionais_Adultos  
1             car purchase  Profissionais_Adultos  
2    purchase of the house  Profissionais_Adultos  
3  supplementary education  Profissionais_Adultos  
4        to have a wedding  Profissionais_Adultos

In [79]:
# Distribuição de `days_employed` medianos com base em seus parâmetros identificados
data_sem_ausente['days_employed'].describe()

count    19239.000000
mean      6819.196092
std       9810.474075
min         24.141633
25%        927.974414
50%       2194.814342
75%       5557.111735
max      27393.750000
Name: days_employed, dtype: float64

Para os dados ausentes na coluna days_employed usaremos a mediana é o elemento do meio. Ela divide nossos dados em duas partes,e podemos substituir o outlier com o seu valor caso necessário.

In [80]:
# Vamos escrever uma função que calcule médias ou medianas (dependendo da sua decisão) com base no seu parâmetro identificado
days_empl_median = 2194.814342
data['days_employed'] = data['days_employed'].fillna(days_empl_median) 

In [81]:
# Verifique se a função funciona
data['days_employed'].isnull().sum()

0

<div class="alert alert-warning">
<strong>Comentários do revisor - v1</strong>

Sempre que possível busque fazer as imputações via código e não escrevendo os números como você fez... A tendência ao erro escrevendo é muito maior.
    
</div>

## Categorização dos dados

In [82]:
# Exiba os valores dos dados selecionados para categorização
data[['family_status','income_type','debt','total_income']].head(5)

family_status income_type  debt  total_income
0            married    employee     0     40620.102
1            married    employee     0     17932.802
2            married    employee     0     23341.752
3            married    employee     0     42820.568
4  civil partnership     retiree     0     25378.572

In [83]:
#Existe alguma relação entre ter filhos e pagar um empréstimo em dia?
#podemos perceber nesta tabela geral que familias entre 1 e 2 filhos são as maiores ofensoras

datagroup = data[['children','debt']].value_counts()
datagroup

children  debt
0         0       13027
1         0        4364
2         0        1858
0         1        1063
1         1         444
3         0         303
2         1         194
4         0          37
3         1          27
5         0           9
4         1           4
dtype: int64

In [84]:
#Existe alguma relação entre o estado civil e o pagamento de um empréstimo no prazo estipulado?
#podemos perceber nesta tabela geral que estado civil como casado é o maior ofensor em débitos

datagroup = data[['family_status','debt']].value_counts()
datagroup

family_status      debt
married            0       11334
civil partnership  0        3748
unmarried          0        2523
divorced           0        1105
married            1         927
widow / widower    0         888
civil partnership  1         385
unmarried          1         273
divorced           1          84
widow / widower    1          63
dtype: int64

In [85]:
# Verificando se existe uma relação entre o nível de renda e o pagamento de um empréstimo no prazo
data['total_income'].describe()

count     21330.000000
mean      26791.503175
std       15667.684517
min        3306.762000
25%       17200.597250
50%       24972.716500
75%       31331.348000
max      362496.645000
Name: total_income, dtype: float64

In [86]:
#função para categorizar nível de renda
def nivel_renda(valor_renda):
    if valor_renda <= 17200:
        return 'baixo'
    if valor_renda >= 17200 and valor_renda <= 23078:
        return 'médio'
    if valor_renda >= 23078 and valor_renda <= 31555:
        return 'alto'
    else:
        return 'muito alto'

In [87]:
#checando se a função funciona corretamente
print(nivel_renda(32333))

muito alto


In [88]:
# Criando a nova coluna que categoriza o nível de renda e verificando se esta correta
data['nivel_renda'] = data['total_income'].apply(nivel_renda)
data.head(5)

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0   27393.750000         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose        Categoria_Idade mediana_income nivel_renda  
0    purchase of the house  Profissionais_Adultos           erro  muito alto  
1             car purchase  Profissionais_Adultos           erro       médio  
2    purchase of the house  Profissionais_Adultos           erro        alto  
3  supplementary education  Profissionais_Adultos           erro  muito alto  
4        to have a wedding  Profissionais_Adultos           erro        alto

<div class="alert alert-success">
<strong>Comentários do revisor - v1</strong>

Bom trabalho!
    
</div>

In [89]:
# Criando a tabela que gera maior visibilidade na categorização idade x renda e débito
# Podemos perceber que profissionais adultos são os maiores ofensores de débitos
pivot_renda = data.pivot_table(index = ['nivel_renda','Categoria_Idade'],
            values = 'debt',
            aggfunc = 'sum',
            )
pivot_renda

debt
nivel_renda Categoria_Idade            
alto        Aposentados              12
            Jovem_Aprendiz            0
            Profissionais_Adultos   555
baixo       Aposentados              14
            Jovem_Aprendiz            2
            Profissionais_Adultos   411
muito alto  Aposentados               5
            Jovem_Aprendiz            0
            Profissionais_Adultos   368
médio       Aposentados               4
            Jovem_Aprendiz            3
            Profissionais_Adultos   358

In [90]:
# Verificando se existe uma relação entre quantidade de filhos e o pagamento de um empréstimo no prazo
# Criando uma função para categorizar a quantidade de filhos x risco

def risk_children(kids):
    if kids > 2:
        return 'high risk'
    if kids >= 1:
        return 'medium'
    else:
        return 'low'

In [91]:
# testando a função
print(risk_children(0))

low


In [92]:
# Criando e verificando a nova coluna referente ao risco criança
data['risco_children'] = data['children'].apply(risk_children)
data.head(5)

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0   27393.750000         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose        Categoria_Idade mediana_income nivel_renda  \
0    purchase of the house  Profissionais_Adultos           erro  muito alto   
1             car purchase  Profissionais_Adultos           erro       médio   
2    purchase of the house  Profissionais_Adultos           erro        alto   
3  supplementary education  Profissionais_Adultos           erro  muito alto   
4        to have a wedding  Profissionais_Adultos           erro        alto   

  risco_children  
0         medium  
1         medium  
2            low  
3      high risk  
4            low

In [93]:
# Verificando estado cívil relacionado aos pogamentos em dia 
datagroup = data[['children','debt']].value_counts()
datagroup

children  debt
0         0       13027
1         0        4364
2         0        1858
0         1        1063
1         1         444
3         0         303
2         1         194
4         0          37
3         1          27
5         0           9
4         1           4
dtype: int64

<div class="alert alert-success">
<strong>Comentários do revisor - v1</strong>

Achei curioso o rótulo de high risk pra mais de dois filhos... Eu não sei se seria tão categórico.
    
Mas os 3 grupos fazem sentido!
</div>

In [94]:
# Agrupando a segmentação family estatus x débitos para estudo
datagroup = data[['family_status','debt']].value_counts()
datagroup

family_status      debt
married            0       11334
civil partnership  0        3748
unmarried          0        2523
divorced           0        1105
married            1         927
widow / widower    0         888
civil partnership  1         385
unmarried          1         273
divorced           1          84
widow / widower    1          63
dtype: int64

Analisando préviamente os dados, podemos identificar que status "married" e "civil parthnership", são as classificações que mais denotam volume de dívidas. 

Abaixo segue uma função para categorizar estes dados, dando maior visibilidade para a segmentação. Inserindo os dados "married" e "civil parthnership" como righ risk e medium.

In [95]:
#função para categorização family Status

def risk_family(family):
    if family == 'married' :
        return 'high risk'
    if family == 'civil partnership':
        return 'medium'
    else:
        return 'low'

In [96]:
print(risk_family('divorced'))

low


In [97]:
data['risco_family_status'] = data['family_status'].apply(risk_family)
data.head(5)

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0   27393.750000         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose        Categoria_Idade mediana_income nivel_renda  \
0    purchase of the house  Profissionais_Adultos           erro  muito alto   
1             car purchase  Profissionais_Adultos           erro       médio   
2    purchase of the house  Profissionais_Adultos           erro        alto   
3  supplementary education  Profissionais_Adultos           erro  muito alto   
4        to have a wedding  Profissionais_Adultos           erro        alto   

  risco_children risco_family_status  
0         medium           high risk  
1         medium           high risk  
2            low           high risk  
3      high risk           high risk  
4            low              medium

In [98]:
data = data[['risco_family_status','debt']].value_counts(normalize = True) * 100
data

risco_family_status  debt
high risk            0       53.136428
low                  0       21.172058
medium               0       17.571496
high risk            1        4.345992
low                  1        1.969058
medium               1        1.804970
dtype: float64

Podemos percebe que temos um grupo de maior risco de 4.34% os casados, 1.96% parceria civil, 1.80 "Outros" em comparativo com o grupo total. 

Percebemos também que também sobre estas segmentações de status familiar são a mais populares dentro do grupo, portanto o volume maior de casados reflete na volumetria de débitos, assim sucessivamente.

<div class="alert alert-danger">
<strong>Comentários do revisor - v1</strong>

Não entendi aqui sua categorização pro estado civil.
    
Sugiro que você:
    * Dê um value_counts() na variável
    * Faça uma análise qualitativa de quais devem ser unidos (tipo casado e união estável) por exemplo
    * Faça as substituições.
    
</div>

<div class="alert alert-danger">
<strong>Comentários do revisor - v1</strong>

Senti falta da categorização do propósito do empréstimo.
    
</div>

Analisando os dados com a segmentação "purpose"

In [99]:
data['purpose'].unique()

KeyError: 'purpose'

In [ ]:
#criando função para categorização de dados 'purpose'
def cond_car(state):
    if state == 'car purchase':
        return('aquisição carro')
    if state == 'buying a second-hand car':
        return('aquisição carro')
    if state == 'buying my own car':
        return('aquisição carro')
    if state == 'cars':
        return('aquisição carro')
    if state == 'to own a car':
        return('aquisição carro')
    if state == 'car':
        return('aquisição carro')
    if state == 'purchase of a car':
        return('aquisição carro')
    if state == 'to buy a car':
        return('aquisição carro')
    if state == 'second-hand car purchase':
        return('aquisição carro')
    if state == 'purchase of the house':
        return('investimento imóvel')
    if state == 'housing transactions':
        return('investimento imóvel')
    if state == 'purchase of the house for my family':
        return('investimento imóvel')
    if state == 'buy residential real estate':
        return('investimento imóvel')
    if state == 'construction of own property':
        return('investimento imóvel')
    if state == 'property':
        return('investimento imóvel')
    if state == 'building a property':
        return('investimento imóvel')
    if state == 'purchase of my own house':
        return('investimento imóvel')
    if state == 'real estate transactions':
        return('investimento imóvel')
    if state == 'buying property for renting out':
        return('investimento imóvel')
    if state == 'housing renovation':
        return('investimento imóvel')
    if state == 'housing':
        return('investimento imóvel')
    if state == 'buy real estate':
        return('investimento imóvel')
    if state == 'buy commercial real estate':
        return('investimento imóvel')
    if state == 'transactions with commercial real estate':
        return('investimento imóvel')
    if state == 'building a real estate':
        return('investimento imóvel')
    if state == 'transactions with my real estate':
        return('investimento imóvel')
    if state == 'supplementary education':
        return('educação')
    if state == 'education':
        return('educação')
    if state == 'to become educated':
        return('educação')
    if state == 'getting an education':
        return('educação')
    if state == 'getting higher education':
        return('educação')
    if state == 'to get a supplementary education':
        return('educação')
    if state == 'going to university':
        return('educação')
    if state == 'getting an education':
        return('educação')
    if state == 'university education':
        return('educação')
    if state == 'to have a wedding':
        return('gasto casamento')
    if state == 'wedding ceremony':
        return('gasto casamento')
    if state == 'having a wedding':
        return('gasto casamento')    
    

In [ ]:
#verificando se a condição funciona 
cond_car('to have a wedding')

In [ ]:
#criando nova coluna para a categorização de 'purpose'
data['purpose'] = data['purpose'].apply(cond_car)
data.head(5)

## Verificar as Hipóteses


In [ ]:
# total de clientes inadimplentes dentro da base de dados (somatória)
data['debt'].sum()

**Nível de Renda**

In [ ]:
data = data[data['debt'] == 1]
data.head()

In [ ]:
data = data.groupby('nivel_renda')['debt'].mean()


In [ ]:
#Existe uma correlação entre o nível de renda e do pagamento em dia?
data[['nivel_renda','debt']].value_counts()

In [ ]:
# inadimplemtes X nivel de renda (nivel renda Muito Alto )
total_muito_alto = round((383 / 1732),2)
total_muito_alto

In [ ]:
# inadimplemtes X nivel de renda (nivel renda Alto )
total_alto = round((467 / 1732),2)
total_alto

In [ ]:
# inadimplemtes X nivel de renda (nivel renda Medio )
total_medio = round((455 / 1732),2)
total_medio

In [ ]:
# inadimplemtes X nivel de renda (nivel renda Baixo )
total_baixo = round((427 / 1732),2)
total_baixo

In [ ]:
# Calculando a média de inadimplencia do grupo
media_nivel_renda = (total_muito_alto + total_alto + total_medio + total_baixo ) / 4 
media_nivel_renda

**Conclusão**

Foi realizado a somatória de todos os clientes que constam como inadimplentes na base de dados e podemos perceber que clientes das duas extremidades, isto é alto e baixa renda são os que possuem mais risco, do que aqueles clientes que são intermediários classificados como médio risco

**Risco Criança**

In [ ]:
# Verifique os dados das crianças e do pagamento em dia
# Calcular a taxa de inadimplência com base no número de filhos

# Verificando o grupo que possuem filhos e agrupando os inadimplentes

data[['risco_children','debt']].value_counts()

In [ ]:
# Inadinplencia calculada em relação ao total de empréstimos realizados no grupo de dados e o risco criança.
total_RNR = 1732
# inadimplencia quantidade criança 3 >
total_RC_alto = round((31 / total_RNR),2)
total_RC_alto

In [ ]:
# inadimplencia quantidade criança 2 e 1 
total_RC_medio = round((638 / total_RNR),2)
total_RC_medio

In [ ]:
# inadimplencia quantidade criança  < 2
total_RC_baixo = round((1063 / total_RNR),2)
total_RC_baixo

In [ ]:
media_risk_children = round((total_RC_alto + total_RC_medio + total_RC_baixo ) / 3,2) 
media_risk_children

**Conclusão**

Dentro do grupo criado para avaliar a quantidade de filhos o que gera uma despesa maior para a familia podemos perceber, que o grupo de familia que possui mais filhos não apresenta um grau de risco maior para realizar os emprestimos e sim o grupo que possui de 1 a 2 filhos (classificados como baixo e médio, nomes que se referem a quantidade de filhos) são o que apresentam maior risco x empréstimo.


**Status Familiar**

In [ ]:
# Verifique os dados de status da família e do pagamento em dia

# Calcular a taxa padrão com base no status da família
data[['risco_family_status','debt']].value_counts()

In [ ]:
total_RF_alto = round((927 / total_RNR),2)
total_RF_alto

In [ ]:
total_RF_medio = round((385 / total_RNR),2)
total_RF_medio

In [ ]:
total_RF_baixo = round((420 / total_RNR),2)
total_RF_baixo

In [ ]:
media_risk_status_f = round((total_RF_alto + total_RF_medio + total_RF_baixo ) / 3,2) 
media_risk_status_f

<div class="alert alert-danger">
<strong>Comentários do revisor - v1</strong>

Podemos melhorar essa sessão.

O uso do value_counts não é o mais adequado para fazermos essa verificação.
O mais interessante seria você aproveitar as categorias que criou anteriormente, agrupar por cada uma delas individualmente utilizando o groupby() e fazer as devidas agregações para avaliar se existe diferença na taxa de inadimplência.


    
</div>

**Conclusão**

Após esta avaliação podemos perceber dentro da base de dados os clientes que foram avaliados como 'alto padrão' são os que possuem maior risco devido a volumetria que os mesmo apresentam obtendo mais débitos, do que os clientes avaliados em médio e baixo padrão.


**Como a finalidade do crédito afeta a taxa de inadimplência?**

O controle da inadimplência são fundamentais para que a empresa evite o comprometimento do fluxo de caixa pelo não recebimento dos clientes nas datas de vencimentos combinados, por isso é fundamental monitorar a inadimplência, para reduzir os seus riscos e problemas com capital de giro.

**Confira os percentuais de inadimplência para cada finalidade de crédito e analise-os**

Dentro desta avaliação tivemos os seguintes resultados referente ao percentual de inadimplência na base de dados seguindo 3 critérios, Nível de Renda, Risco Criança e Status Familiar. 

A média geral para estes 3 grupo é de 30,33%

Tivemos o retorno da seguinte forma percentual para cada grupo> 

- Nível de Renda - InadimpLência 25% (Abaixo da média do grupo em -5,33pp)
- Risco Criança - InadimpLência 33% (Acima da média do grupo em +3pp)
- Status Familiar - InadimpLência 33% (Acima da média do grupo em +3pp)

Abaixo segue o detalhamento por subgrupos avaliados em Alto, Médio e Baixo:



**Dados percentuais de inadimplência detalhados para cada grupo de classificação usado para realizar o estudo.**

**Nível de Renda**

 **Alta Renda**
    
Dentro da avaliação realizada a média percentual de inadimplência do grupo é de 25% e estudando o nível de renda dos cliente percebemos um pequeno desvio no comportamento dos clientes, que possuem um nível de renda maior de +3.pp (28%) na assiduidade do pagamento dos seus débitos, do que os clientes de média e baixa renda, neste caso baseado em dados estamos indo ao contrário do pensamento que cliente "mais ricos" são bons pagadores.
    
 **Média Renda** 
    
No estudo realizado do grupo de clientes que apresentaram risco médio podemos destacar que é o "melhor grupo" no sentido que possui o menor percentual de risco ficando com -3pp. abaixo da média geral do grupo.
    
 **Baixa Renda**
    
Para este grupo esta dentro da média de 25%, não apresentando desvio em relação aos demais grupos avaliados como alto e média.
    
**Risco Criança**

Para este grupo ao analisar a quantidade de filhos x risco crédito, sendo clientes que já possuem dívidas, adotamos o critério que quanto maior o número de filhos maior seria o risco, portando as nomenclaturas utilizadas, foram famílias que possuem de 3 a mais filhos como alto risco, 1 e 2 filhos médio e 0 filhos (com dívida) baixo. 

Nesta avaliação tivemos um desvio considerável ao verificar que a média do grupo ficou em 33%, sendo que famílias que não possuem são as mais ofensoras nos resultados, abaixo segue detalhamento:

  
**Maior quantidade de filhos**
    
Famílias que possuem um número maior que filhos ( de 3 ou +) e que geram um maior gasto no seu orçamento familiar, ficaram com 0,02% de cálculo de inadimplência e seu comportamento referente ao grupo foi de -31pp em relação a media do grupo (33%), sendo este grupo aquele que oferece menor risco. 
    
**Quantidade de filhos média**
    
O grupo intermédiário de famílias que possuem 1 a 2 filhos, ficaram com um total de 37% sendo +4pp acima da média do grupo. 
    
**Sem filhos**
    
Este grupo contém familias que não possuem filhos e seu resultado ficou em 61% o maior de todos os demais, sendo +28pp a mais da média do grupo, sendo o maior ofensor.    

**Status Familiar**

    
Para este quesito avaliamos este grupo em 3 estágio, sendo "Alto Padrão", "Médio" e "Baixo", este cálculo foi baseado na mediana do grupo em relação aos seus recebimentos, onde foi usado um cálculo descritivo para segmentar o grupo em seus valores. 
    
E o resultado da média deste 3 grupos(alto,médio e baixo), ficou em 33%.
    
Detalhamento segue abaixo:
    
**Alto Padrão**
    
Ao realizar a verificação dentro do grupo de "alto padrão", possuímos um risco de +21pp acima da média do grupo, mostrando uma cartela de clientes mais ofensora a liberação de crédito, o que entra em contra partida de serem clientes com maior "poder aquisitivo demostrando nos seus ganhos comparados aos demais do grupo".
    
**Médio**
    
Dentro do grupo de clientes intermediários o resultado apresentou 22% de risco inadimplência o menor do grupo ficando -11pp abaixo da média do grupo.
        
**Baixo**
    
Dentro deste grupo ficou com -9pp abaixo da média do grupo.
    



<div class="alert alert-success">
<strong>Comentários do revisor - v1</strong>

Ótimo trabalho com as conclusões.

Sua análise está bem legal e bem completa, os detalhes que apontei acima vão te ajudar a fazer essas análises com mais agilidade.
    
</div>

# Conclusão Geral 

Com base nos dados e manipulações realizadas este estudo foi dirigido para primeiramente analisar a base de dados realizar verificações de possíveis ausência, corrompimento e/ou demais variáveis que pudessem interferir na tratativa até a chegada da construção e/ou normalização de dados e gerar informações e análises sobre o contexto de empréstimo x análise de crédito da carteira de clientes informada.  

Podemos destacar algumas das ações realizadas da seguinte forma:

Foram identificados dados simétricos em duas colunas 'days_employed' e 'total_income', onde foram realizada as tratativas de inserção de valores basiados em cálculos de medianas, afim de não aferir a integridade dos cálculos.

Normalização de título de dados (categóricos) ao que se refere sua digitação ser computada da forma correta para avaliação de dados exemplo coluna 'education' e normalização de dados negativos na coluna 'days_employed'(numéricos). 

Realizada a exclusão de dados duplicados, afim de não aferir resultados e/ou análises futuras. 

Ao verificar já no início da tratativa os dados ausentes não eram um volume comprometedor na base de dados abaixo segue informação:

- Total nulos,  2174
- Total de dados no geral,  21525
- Temos um total percentual de dados ausentes na base de  0.1 %

É perceptível que o percetual fica abaixo de 1% do total de dados, porém decidimos não excluir os dados e sim realizar o preenchimento com cálculos simétricos das próprias colunas de dados ausentes.

Realizado uma triagem na verificação dos dados até a escolha de dados que embassaram este estudo, verificado gênero dos clientes, tipo classificação profissional, educação, idade entre outros.

Criado uma categorização de dados para as seguintes informações, status familia, risco crianças e renda familiar, para seguimentar subgrupos e fazer abertura de dados para maiores verificações. 

**Considerações Finais**

Concluímos que o grupo de Risco Criança e Status Familia, são os maiores ofensores ao risco a liberação de crédito no valor de 33% de risco de inadimplência. 

Em Risco Criança verificamos uma abertura no perfil do grupo que famílias que possuem menos crianças possuem mais inadimplência, no que o grupo que possuem 3 ou mais filhos, demonstrando que não é a quantidade de filhos que afere o risco. 

Para Status Familia o indicador é puxado para 33% devido ao grande desvio de +21pp de clientes que possuem um alto padrão de renda, porém o perfil é de maior inadimplente.

Renda Familiar assim como Status familiar o comportamento se reproduziu em renda familiar confirmando a hipótese acima. 






<div class="alert alert-info">
<strong>Comentário do revisor v1</strong>

Bom trabalho até aqui!

Identifiquei alguns pontos de melhoria que gostaria que você desse um pouco de atenção....
    
Por favor foque em:
    
* Dar um double check nos itens apontados na primeira sessão. O resultado foi bom, mas ficaria mais legal se detalhasse melhor o seu processo decisório.
* Podemos trabalhar melhor a imputação na variável children do que simplesmente eliminar as linhas
* Há outros problemas com a variável days_employed
* Há um erro (provavelmente de continuidade) no notebook que não permite a conclusão da revisão.
    
O mais crítico foi o erro de código que impediu que eu concluísse a análise, mas você corrigindo nós seguimos.

    
Dê uma olhada nesses pontos e ai continuamos de onde paramos, pode ser?


</div>

<div class="alert alert-info">
<strong>Comentário do revisor v2</strong>

Muito obrigado pelas correções que você fez.
    
Está no caminho certo e dá pra ver que colocou bastante esforço no trabalho que realizou.
    
Gostaria apenas de uma última interação sua para chegarmos no resultado definitivo desse trabalho, pode ser?
    
Gostaria que você desse uma olhada:
    
* nos pontos que anotei na parte de categorização
    
* tomar cuidado com atribuições 'hard coded' ou seja, atribuindo um número a força vs por meio de código ( exemplo x= 1234 vs x = y.mean(), ainda que sejam em tese equivalentes, é muito frequente incorrer em erros no primeiro método)
*Na verificação das hipóteses fazer agrupamentos e agregações para maior agilidade e visibilidade dos resultados.
    
Estamos quase lá! Até breve.

</div>